[cbuild-pack.yml] Document cbuild-pack.yml

slhultgren opened the pull request:
Document cbuild-pack.yml introduced in PR https://github.com/Open-CMSIS-Pack/devtools/pull/1143

Fixes https://github.com/Open-CMSIS-Pack/devtools/issues/1122

Contributed by STMicroelectronics

Easiest way to read it is by looking at the rendered form https://github.com/slhultgren/cmsis-toolbox/blob/pr/shu-cbuild-pack-documentation/docs/YML-CBuild-Format.md#pack-locking

brondani commented on docs/YML-CBuild-Format.md:
````diff
@@ -33,13 +33,47 @@ File                | Description
+        - ARM::CMSIS@5.9.0
+```
+
+If a context is added or changed, the `selected-by` is used to ensure that the `pack requirements` are resolved to a `consistent` pack version, reducing surprising versions being selected in the entire application.
````
I think the word "changed" in this sentence can be misleading given the fact the approach is additive: new packs (regardless of being added into new or existing contexts) are also added to `cbuild-pack`, but if existing packs are removed from input files they are not automatically removed from `cbuild-pack`. In such case there will be potentially dangling entries so a clean-up mechanism should be also provided/recommended/documented - what other package managers do in this case?

slhultgren replied:
Both `yarn` and `cargo` cleans the lockfile in this case, I will add a note about this in the documentation. The implementation PR needs to be updated slightly.

mcgordonite commented on docs/YML-CBuild-Format.md:
```diff
@@ -33,13 +33,47 @@ File                | Description
+- It contains strict list of software packs used including version information and ranges from `.csolution.yml` and all subsequent potentially included `.cproject.yml` and `.clayer.yml` files.
+- This file is used as `lock-file` to ensure that subsequent runs of `csolution` use the very same packs and pack versions on all computers / setups / runs / environments.
+
+### Pack locking
```
Is it possible to document when the lock file is updated in the user workflow?

jkrech replied:
@slhultgren, I have not looked at the source code, but my expectation would be that the update of the `<solution>.cbuild-pack.yml` would also be included in the `update-rte` command but not in any of the `list` commands.

slhultgren replied:
Clarified this in the updated docs

slhultgren commented:
Updated the PR with a bit more clarification on when the cbuild-pack.yml is used and updated:
Easiest to read is here: https://github.com/slhultgren/cmsis-toolbox/blob/pr/shu-cbuild-pack-documentation/docs/YML-CBuild-Format.md#pack-locking

jkrech commented:
@slhultgren  thanks a lot for adding detail to the documentation regarding the reading and writing of the cbuild-pack.yml file and which `csolution commands` are affected, as well as the maintenance of entries.
Could you please confirm that the cbuild-pack.yml does not contain any kind of history information anymore and therefore there is no need for a command for cleaning up.

We are frequently discussing the aspect of "backward compatibility". With the introduction of the cbuild-pack.yml the previous behavior is changed such that after installing an updated pack version used by a csolution which does not fix the pack version, the newly installed pack version will not be used.

My request is to add a `command line option` for both cbuild and csolution that allows to enable/disable the use of cbuild-pack.yml.

For strict backward compatibility the cbuild-pack.yml would be disabled by default and require 
```
- P --pack-lock-on            Lock pack versions using *.cbuild-pack.yml
```

We could consider an explicit entry in the `csolution.yml` file similar to enabling the use of `cdefault.yml`
```
solution:
  cdefault:
  cbuild-pack:
  ...
```

ReinhardKeil commented:
@slhultgren thanks for documenting the feature.  

I agree with @jkrech to add a option for enabling this feature.

mcgordonite commented on docs/YML-CBuild-Format.md:
Extending @jkrech's suggested `-P` flag in https://github.com/Open-CMSIS-Pack/cmsis-toolbox/pull/76#issuecomment-1805688537, it would be useful to add a `--frozen-lockfile` flag for CI workflows. When this is set, all commands should fail if the lockfile is not up to date with the cproject input files.

See similar in the yarn package manager: https://classic.yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-frozen-lockfile.

jkrech replied:
Command line options will be added separately.

ReinhardKeil commented:
I believe just one option is enough.  By default the file `*.cbuild-pack.yml` is written and records all packs including versions that are used by the project files (*.csolution.yml, *.cproject.yml, *.clayer.yml.

Option:
`-P, --pack-lock`      Use `*.cbuild-pack.yml` as input to lock software pack usage.

The option `--pack-lock` reports an error when the file `*.cbuild-pack.yml` does not exist, or when there are different packs requested by the project files.

brondani commented on docs/YML-CBuild-Format.md:
```diff
@@ -33,13 +33,87 @@ File                | Description
+
+subgraph Loading
+  res1["<center><u>Resolving step 1 (PopulateContexts)</u></center><ul><li>Reading the files</li><li>Adds cbuild-pack.yml\npack requirements to the solution</li></ul>"]
+  res2["<center><u>Resolving step 2 (AddPackRequirements)</u></center><ul><li>Pack version ranges are matched to fully qualified versions using cbuild-pack.yml</li><li>Pack wildcards are matched to fully qualified versions using cbuild-pack.yml</li><li>Pack wildcards are kept for further expansion in the model</li><li>Local packs are left as-is</li></ul>"]
```
What does "Pack wildcards are kept for further expansion in the model" mean?
After resolving the packs, no further processing should be needed.

jkrech reviewed (approved):
Looks good to me.
Extensions like command line options to be added in separate PR.

jkrech commented on docs/YML-CBuild-Format.md:
````diff
@@ -33,13 +33,87 @@ File                | Description
+```
+
+If a context is added or changed, the `selected-by` is used to ensure that the `pack requirements` are resolved to a `consistent` pack version, reducing surprising versions being selected in the entire application.
+If a `pack requirement` is no longer present in the application, then the `cbuild-pack.yml` file will be *cleaned* from the relevant entries during the `cbuild-pack.yml`-generation step.
````
clean will be done for all contexts only.
